File: python/year_2015/day_05.py
"""
https://adventofcode.com/2015/day/5
"""

from itertools import islice, starmap
from operator import eq, ne
from pathlib import Path

from iteration_utilities import successive


def part_1(text):
    def nice(string):
        if any(two in string for two in ("ab", "cd", "pq", "xy")):
            return False

        if all(starmap(ne, successive(string))):
            return False

        return 3 <= sum(map(string.count, "aeiou"))

    return sum(map(nice, text.split()))


def part_2(text):
    def split_pair(string, start):
        return any(starmap(eq, successive(islice(string, start, None, 2))))

    def twin_pair(first, second):
        if second in seen:
            return True
        seen.add(first)

    def nice(string):
        if not split_pair(string, 0) and not split_pair(string, 1):
            return False

        seen.clear()

        return any(starmap(twin_pair, successive(successive(string))))

    seen = set()

    return sum(map(nice, text.split()))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_05.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 238


def test_part_1_example_1():
    assert part_1("ugknbfddgicrmopn") == 1


def test_part_1_example_2():
    assert part_1("aaa") == 1


def test_part_1_example_3():
    assert part_1("jchzalrnumimnmhp") == 0


def test_part_1_example_4():
    assert part_1("haegwjzuvuyypxyu") == 0


def test_part_1_example_5():
    assert part_1("dvszwmarrgswjxmb") == 0


def test_part_2_input():
    assert part_2(_read_input()) == 69


def test_part_2_example_1():
    assert part_2("qjhvhtzxzqqjkmpb") == 1


def test_part_2_example_2():
    assert part_2("xxyxx") == 1


def test_part_2_example_3():
    assert part_2("uurcxstgmygtbstg") == 0


def test_part_2_example_4():
    assert part_2("ieodomkazucvgmuy") == 0

File: rust/year_2015/src/day_05.rs
use itertools::Itertools as _;

pub fn part_1(input: &str) -> usize {
    input
        .lines()
        .filter(|line| {
            let bytes = line.as_bytes();
            let vowel_count = bytes
                .iter()
                .filter(|b| matches!(b, b'a' | b'e' | b'i' | b'o' | b'u'))
                .count();
            let has_double = bytes.array_windows().any(|[a, b]| a == b);
            let has_forbidden = bytes
                .array_windows()
                .any(|pair| matches!(pair, b"ab" | b"cd" | b"pq" | b"xy"));
            vowel_count >= 3 && has_double && !has_forbidden
        })
        .count()
}

pub fn part_2(input: &str) -> usize {
    input
        .lines()
        .filter(|line| {
            let bytes = line.as_bytes();
            let has_pair = bytes
                .array_windows::<2>()
                .enumerate()
                .any(|(i, pair)| bytes[i + 2..].array_windows().contains(pair));
            let has_repeat = bytes.array_windows().any(|[a, _, c]| a == c);
            has_pair && has_repeat
        })
        .count()
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_05.txt");

    #[test_case("ugknbfddgicrmopn" => 1)]
    #[test_case("aaa" => 1)]
    #[test_case("jchzalrnumimnmhp" => 0)]
    #[test_case("haegwjzuvuyypxyu" => 0)]
    #[test_case("dvszwmarrgswjxmb" => 0)]
    #[test_case(INPUT => 238)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    #[test_case("qjhvhtzxzqqjkmpb" => 1)]
    #[test_case("xxyxx" => 1)]
    #[test_case("uurcxstgmygtbstg" => 0)]
    #[test_case("ieodomkazucvgmuy" => 0)]
    #[test_case(INPUT => 69)]
    fn part_2(input: &str) -> usize {
        super::part_2(input)
    }
}
